### DISCUSSION TITLE
[RESEARCH] On board Communication

### DISCUSSION THREAD

**Discussion by pablo-collantes** [Comet-Robotics/URC#10] [Networking]
On our rover we eventually want to connect all our components through ethernet to our main computer.
By using ethernet we can use a router/switch that can keep everything on the same network making it easy to communicate and making it so devices on and off the rover have access to every ROS2 node in the framework.

Before we can implement this we need to research specifically what protocol we would need to use or which would be easiest to create.
Research:
Ethernet Protocols
Why have some people chosen [EtherCAT](https://dewesoft.com/blog/what-is-ethercat-protocol)

**Comment by CrispyAlice2** [comment]
The communications link is most likely going to be lossy at certain points, and because of this TCP's head of line blocking and ack delay is going to kill the data rate. And, the proposed ham band link it will be slow enough that avoiding the tcp/ip overhead would be needed. It might be worth looking into having at least some things (especially essential command and control) using a custom protocol on top of udp / any generic datagram protocol that could go over either link. Whatever protocol is chosen here doesn't have to be used by the actual ros2 nodes, there could be a box on the rover that handles this protocol and turns it into ros2 commands. This could also help shave down unnecessary verbosity in the communications link, we could have a super well trimmed protocol with little overhead to maximally fit in the link.

**Reply by conner-replogle** [reply]
If we cannot just simply do the P2P networking and have it be like a normal LAN network to interact with the ROS2 nodes. We will most likely implement a ROS2 Node that is connected with the antenna and its whole job is sending the required data.

**Comment by conner-replogle** [comment]
Ignoring the radio aspect of things we urgently need to buy the router for the rover so we can test the connection between bby pc and motor board I'm thinking we just snipe a cheap wifi router

**Reply by CrispyAlice2** [reply]
I have an old router/switch/access point combo unit we can borrow until we get a switch and whatnot.

**Reply by conner-replogle** [reply]
Can we just use a switch I would imagine it can't assign an ip without a router

**Reply by Kandles11** [reply]
You could always run a DHCP server on the mini PC, or more simply, just set static IP's for the devices manually

**Reply by CrispyAlice2** [reply]
It probably should be static IPs anyway, that way each node is always the same address.

**Comment by conner-replogle** [comment]
Okay for now I think my switch will work then

**Comment by CrispyAlice2** [comment]
https://github.com/epeters13/pyLoraRFM9x
